### ISSUE TITLE
Stop using keys

### ISSUE THREAD

**Issue by ghost** [rv-kip/express-redis-cache#55]
Hey there,

In our production environment this was pretty slow (https://s.put.re/4ueAZEA.png).
You can read more details about .keys under: http://redis.io/commands/keys

On this screenshot you can see our t2.small redis cache instance cpu usage on amazon, take a look at 2:50. That's where I replaced your caching lib with a custom one that uses .get (https://s.put.re/f36CV4z.png).

Maybe your usage of .keys has a deeper point I don't see here, but let me quote the redis docs:

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases. This command is intended for debugging and special operations, such as changing your keyspace layout. Don't use KEYS in your regular application code. 

greetz

**Comment by JADurham909** [comment]
Second this, we had major problems from the keys call in get.js.  I have since reworked it to use scan in its place.  It was blocking the entire cache server.

**Comment by rv-kip** [comment]
I agree with this. The use of `keys` predates my stewardship and it obviously has performance drawbacks. The concern I have is backwards compatibility if users expect  a request to `foo` actually means would return results `foo1`, `foo2`, etc.

There is the expectation of wildcards
https://github.com/rv-kip/express-redis-cache/blob/master/lib/ExpressRedisCache/get.js#L30

**Comment by JADurham909** [comment]
Scan will do exactly the same function without blocking the sever with newer versions of redis.  I can post a PR when my changes are done in my fork if you wish.

http://redis.io/commands/scan

Basically you just recursively scan until all keys are retrieved and wrap it in a function to replace keys.

**Comment by rv-kip** [comment]
Please do! Thanks @JADurham909.

**Comment by bsyk** [comment]
Could the get.js file handle wildcards in the same way as del.js?
Most of the get calls in my app are for single keys.  More than 50% of the time spent in Redis (according to NewRelic) is doing the keys call.
https://github.com/rv-kip/express-redis-cache/blob/master/lib/ExpressRedisCache/del.js#L44

```
      /** Detect wilcard syntax */
      var hasWildcard = redisKey.indexOf('*') >= 0;

      /** If has wildcard */
      if ( hasWildcard ) {
        /** Get a list of keys using the wildcard */
```

**Comment by Yurko-Fedoriv** [comment]
Seems like we got downtime from this. Our redis choked and it's slowlog shows incredible amount of KEYS calls.

**Comment by Yurko-Fedoriv** [comment]
And this is how redis responded to replacing the original lib with fork from @bsyk 
Thanks, Ben.

![image](https://cloud.githubusercontent.com/assets/1327761/19733008/5f36f432-9bab-11e6-97f3-9060bb868506.png)

**Comment by bsyk** [comment]
You're very welcome. 👍
We had the same issue and results.

**Comment by wellcaffeinated** [comment]
+1

**Comment by Yurko-Fedoriv** [comment]
@bsyk you might want to check out #85 because you ported it in your fork

**Comment by bsyk** [comment]
@Yurko-Fedoriv has the EventEmitter issue been fixed in master?  I updated my fork a few days ago.

**Comment by Yurko-Fedoriv** [comment]
@bsyk nope, it's still there https://github.com/rv-kip/express-redis-cache/blob/master/lib/ExpressRedisCache/route.js#L94-L96
I had to quickly [hotfix](https://github.com/nutritionix/express-redis-cache/blob/avoid_keys_call/lib/ExpressRedisCache/route.js#L94-L96) it for myself, because it already caused us couple downtimes from servers eating up all the memory and becoming non-responsive.

Just letting you know, since I was originally using your fork.

**Comment by bsyk** [comment]
@Yurko-Fedoriv got it. Sorry for causing you issues. If you have a work around for now, I won't update the fork, but I'll keep a close eye on the PR for #85 as well as the PR for #55

**Comment by Yurko-Fedoriv** [comment]
@bsyk no worries, just wanted to let you know before it also hits you. I'm good for now.
Thanks again for fix to this one.

**Comment by robdvr** [comment]
You are welcome to use our fork until these PRs are reviewed and merged. In your `package.json`:
`"express-redis-cache": "git://github.com/SpringboardAuto/express-redis-cache#master"`

It includes:
- Memory leak fix #85 #86 (Thanks to @LiorRabin)
- Performance enhancements and other fixes #70 #55 #68 #56 (Thanks to @bsyk)
- Added default HTTP Status, and return the original response code (Thanks to @ahnkee)

Hope this helps someone!

**Comment by bharatzvm** [comment]
It does
